refactor: rename `test/utils` -> `test/internal`

Tal-or opened the pull request:
Force the convention that test code cannot be used outside of `test` package.

In addition allow running unit-tests for files under `test/internal` because there is a test code which is complex and/or being used wide enough that it's worth running unit-tests on it.

Tal-or committed:
tests: rename test/utils -> test/internal

renaming to emphasis the tools under tests are internal only and
should be used solely by test code.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
makefile: run unit tests for `test/internal`

There is a test code which is complex enough and used wide enough
so it worth running unit-tests for it.

Changing the makefile so it will consider unit-tests under `test/internal`
as well.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
readme: add readme file

Explain the purpose `test/internal` and its alternatives.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/hold

depends on #1169

Tal-or commented:
/hold cancel

swatisehgal commented on test/internal/README.md:
```diff
@@ -0,0 +1,6 @@
+Code here (test/internal) meant to be used only by test code.
```
````suggestion
Code here (test/internal) is meant to be used only by test code.
````

swatisehgal commented on test/internal/README.md:
```diff
@@ -0,0 +1,6 @@
+Code here (test/internal) meant to be used only by test code.
+
+Please think twice before add code here.
```
````suggestion
Please think twice before adding code here.
````

swatisehgal reviewed (commented):
Some NITs, but overall LGTM

/lgtm

Tal-or commented:
@swatisehgal Thank you for the review I fixed the typo, please approve again when possible

swatisehgal reviewed (commented):
/lgtm

ffromani commented on Makefile:
```diff
@@ -118,7 +118,7 @@ test-unit: test-unit-pkgs test-controllers
 .PHONY: test-unit-pkgs
 test-unit-pkgs: generate-source
-	go test $$(go list ./... | grep -vE 'controller|test|tools|cmd')
+	go test $$(go list ./... | grep -vE 'controller|test/e2e|test/deviceplugin|tools|cmd')
```
suggestion:
```
go test $( go list ./... | grep -vE 'tools|cmd|internal/controller|test/e2e|test/deviceplugin|k8simported' )
```

Tal-or replied:
Done

ffromani commented:
/hold

PTAL to the inline comments

ffromani commented:
/hold cancel
/lgtm

ffromani commented:
/lgtm cancel

linter lane failures

Tal-or commented:
> /lgtm cancel
> 
> linter lane failures

Fixed.
Not sure what has changed, I don't see the errors when running against the main branch

ffromani commented:
> > /lgtm cancel
> > linter lane failures
> 
> Fixed. Not sure what has changed, I don't see the errors when running against the main branch

Previously that package was ignored by the linter, we need to update the linter config. That package was copy/pasted from upstream kubernetes, better to keep it 1:1 unchanged. IOW we either update our local clone or fix the linter config to ignore it. Or both.

Tal-or commented:
> > > /lgtm cancel
> > > linter lane failures
> > 
> > 
> > Fixed. Not sure what has changed, I don't see the errors when running against the main branch
> 
> Previously that package was ignored by the linter, we need to update the linter config. That package was copy/pasted from upstream kubernetes, better to keep it 1:1 unchanged. IOW we either update our local clone or fix the linter config to ignore it. Or both.

Even the most updated package from k8s contains the same lint error.
I don't have a strong opinion, we may set the linter to ignore, or update the local clone. Whatever works the best

ffromani commented:
> > > > /lgtm cancel
> > > > linter lane failures
> > > 
> > > 
> > > Fixed. Not sure what has changed, I don't see the errors when running against the main branch
> > 
> > 
> > Previously that package was ignored by the linter, we need to update the linter config. That package was copy/pasted from upstream kubernetes, better to keep it 1:1 unchanged. IOW we either update our local clone or fix the linter config to ignore it. Or both.
> 
> Even the most updated package from k8s contains the same lint error. I don't have a strong opinion, we may set the linter to ignore, or update the local clone. Whatever works the best

let's update the linter config to update that specific part then, and we move on.

Tal-or committed:
lint: exclude k8simports

This dir contains helper which were copied verbatim from k/k repo.
we prefer to keep the match 1:1 than fixing the linter issues, so
let exclude this directoy from the lint config.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani commented:
/lgtm

ffromani commented:
/retest

ffromani commented:
/test ci-e2e-install-hypershift
